gtk/TorrentCellRenderer.cc

```diff
@@ -170,151 +143,161 @@ Glib::ustring getShortTransferString(
 
     default:
+        return {};
+    }
+}
+
+static std::optional<std::string> getErrorString(tr_stat const* st)
+{
+    switch (st->error)
+    {
+    case TR_STAT_TRACKER_WARNING:
+        return fmt::format(_("Tracker warning: '{warning}'"), fmt::arg("warning", st->errorString));
         break;
+
+    case TR_STAT_TRACKER_ERROR:
+        return fmt::format(_("Error: '{error}'"), fmt::arg("error", st->errorString));
+        break;
+
+    case TR_STAT_LOCAL_ERROR:
+        return fmt::format(_("Local error: '{error}'"), fmt::arg("error", st->errorString));
+
+    default:
+        return std::nullopt;
     }
+}
 
-    return gstr;
+static auto getActivityString(
+    tr_torrent const* const tor,
+    tr_stat const* const st,
+    double const uploadSpeed_KBps,
+    double const downloadSpeed_KBps)
+{
+    switch (st->activity)
+    {
+    case TR_STATUS_STOPPED:
+    case TR_STATUS_CHECK_WAIT:
+    case TR_STATUS_CHECK:
+    case TR_STATUS_DOWNLOAD_WAIT:
+    case TR_STATUS_SEED_WAIT:
+        return getShortStatusString(tor, st, uploadSpeed_KBps, downloadSpeed_KBps);
+
+    case TR_STATUS_DOWNLOAD:
+        if (!tr_torrentHasMetadata(tor))
+        {
+            return fmt::format(
+                ngettext(
+                    "Downloading metadata from {active_count} connected peer ({percent_done:d}% done)",
+                    "Downloading metadata from {active_count} connected peers ({percent_done:d}% done)",
+                    st->peersConnected),
+                fmt::arg("active_count", st->peersConnected),
+                fmt::arg("percent_done", 100.0 * st->metadataPercentComplete));
+        }
+
+        if (st->peersSendingToUs != 0 && st->webseedsSendingToUs != 0)
+        {
+            return fmt::format(
+                ngettext(
+                    "Downloading from {active_count} of {connected_count} connected peer and webseed",
+                    "Downloading from {active_count} of {connected_count} connected peers and webseeds",
+                    st->peersSendingToUs + st->webseedsSendingToUs),
+                fmt::arg("active_count", st->peersSendingToUs + st->webseedsSendingToUs),
+                fmt::arg("connected_count", st->peersConnected + st->webseedsSendingToUs));
+        }
+
+        if (st->webseedsSendingToUs != 0)
+        {
+            return fmt::format(
+                ngettext(
+                    "Downloading from {active_count} of webseed",
+                    "Downloading from {active_count} of webseeds",
+                    st->webseedsSendingToUs),
+                fmt::arg("active_count", st->webseedsSendingToUs));
+        }
+
+        return fmt::format(
+            ngettext(
+                "Downloading from {active_count} of {connected_count} connected peer",
+                "Downloading from {active_count} of {connected_count} connected peers",
+                st->peersSendingToUs),
```

Comment:
Pluralized against `connected_count` (total), not `active_count` (part).
```suggestion
                st->peersConnected),
```

Reply:
66d0b22493eac433bef0895e568f903e231e2880